### PR TITLE
Taxonomy Localized Field

### PR DESCRIPTION
**Pull request by urbanit**
- [x] Allow Taxonomies Localized Version
- [x] Partially solves #4845

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Taxonomies/OrchardCore.Taxonomies.csproj [hunk 1]
```diff
@@ -23,6 +23,7 @@
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Liquid.Abstractions\OrchardCore.Liquid.Abstractions.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Navigation.Core\OrchardCore.Navigation.Core.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.ResourceManagement\OrchardCore.ResourceManagement.csproj" />
+    <ProjectReference Include="..\OrchardCore.ContentLocalization\OrchardCore.ContentLocalization.csproj" />
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Taxonomies/OrchardCore.Taxonomies.csproj:26]: Why you didn't use `ContentLocalization.Abstraction`?

Comment (urbanit) [reply src/OrchardCore.Modules/OrchardCore.Taxonomies/OrchardCore.Taxonomies.csproj:26]: I need ContentLocalization for _localizationEntries... No? Btw, what's the gain of Abstractions?

File: src/OrchardCore.Modules/OrchardCore.Taxonomies/Liquid/TaxonomyTermFilter.cs [hunk 2]
```diff
@@ -38,6 +51,7 @@ public async ValueTask<FluidValue> ProcessAsync(FluidValue input, FilterArgument
             {
                 taxonomyContentItemId = jobj["TaxonomyContentItemId"].Value<string>();
                 termContentItemIds = ((JArray)jobj["TermContentItemIds"]).Values<string>().ToArray();
+                localizedCulture = jobj.Root["LocalizationPart"]["Culture"].Value<string>();
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Taxonomies/Liquid/TaxonomyTermFilter.cs:54]: There should be a LocalizedAspect for that. This way you don't depend on a specific part, and it's strongly typed.

Comment (urbanit) [reply src/OrchardCore.Modules/OrchardCore.Taxonomies/Liquid/TaxonomyTermFilter.cs:54]: Not sure if there is, I couldn't get it. Any hint, appreciated..

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Taxonomies/Liquid/TaxonomyTermFilter.cs:54]: > There should be a LocalizedAspect for that.

I think Seb means `CultureAspect`

**Comment by deanmarcussen** [comment]
Nice one @urbanit this looks like a good start.

Sebastien suggests we should have a `LocalizationAspect` but right now I don't think we have one, so you would have to implement it.

For an example of how aspects work look at the https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Modules/OrchardCore.Html/Handlers/HtmlBodyPartHandler.cs for the `BodyAspect`

The idea would be that the `LocalizationAspect` is in the `ContentLocalization.Abstractions` project and the handler implementation in the `ContentLocalization` module.

This way you don't have to reference the module (which is bad!)

I would imagine that you need the same thing for the `LocalizationEntries`. A `LocalizationEntryAspect` which will provide the entries, if the `ContentLocalization` module is activated.

Right now, with this code, unless the `ContentLocalization` module is activiated, you will get a null reference exception when using taxonomies, because the service is not registered.

If we didn't use an aspect for this, we would need a `NullLocalizationEntries` service, registered by default.

But I think another aspect will work well.

This begs the question then if you need the `Localized` setting on the field.
If you removed it, then any taxonomy that was localized, would automatically localize it's fields. (if the aspect provided the entries), and if it wasn't localized the aspect would provide no entries, because the handler would not run.

So that's a question, rather than a request, because you know more about the requirements, but I think it makes sense, that it happens automatically, if you localize the taxonomy?

You might also want to consider the same feature for the `Tags` driver.

**Comment by hishamco** [comment]
@urbanit if you need help, I can handle the localization parts, `LocalizationAspect`, `NullLocalizationEntries` .. etc and leave the remaining things to you ;)

**Comment by urbanit** [comment]
@deanmarcussen thanks for the feedback. You are right that ContentLocalization should not be enabled in order Taxonomies to run smoothly. So, Aspect (now it;s clear how it works) is best approach. 

@hishamco any help is appreciated! If you have time and mood, please go ahead.

About localized and tags: it could be cases that we do not want to localize taxonomies. Tags is one of them... For a tag, maybe it could be user decision to show all related content items or current language only.

File: src/OrchardCore.Modules/OrchardCore.Taxonomies/OrchardCore.Taxonomies.csproj (hunk 1 above)
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Taxonomies/OrchardCore.Taxonomies.csproj:26]: We don't reference module implementation from another module, I think @deanmarcussen point to that below

**Comment by hishamco** [comment]
> Sebastien suggests we should have a LocalizationAspect but right now I don't think we have one, so you would have to implement it.

@deanmarcussen why we didn't use `CultureAspect`?

**Comment by sebastienros** [comment]
Yeah, CultureAspect is that is the name that we have, sorry

**Pull request closed without merging**